// vuestic-ui/dist/web-components/src/composables/useForm/useForm.js
import { ref, computed } from "vue";
import { u as useTemplateRef } from "../useTemplateRef.js";
const useForm = (elRef) => {
  const form = typeof elRef === "string" ? useTemplateRef(elRef) : typeof elRef === "undefined" ? ref() : elRef;
  return {
    formRef: form,
    isValid: computed(() => {
      var _a;
      return ((_a = form.value) == null ? void 0 : _a.isValid) || false;
    }),
    immediate: computed(() => {
      var _a;
      return ((_a = form.value) == null ? void 0 : _a.immediate) || false;
    }),
    isLoading: computed(() => {
      var _a;
      return ((_a = form.value) == null ? void 0 : _a.isLoading) || false;
    }),
    isDirty: computed(() => {
      var _a;
      return ((_a = form.value) == null ? void 0 : _a.isDirty) || false;
    }),
    isTouched: computed(() => {
      var _a;
      return ((_a = form.value) == null ? void 0 : _a.isTouched) || false;
    }),
    fields: computed(() => {
      var _a;
      return ((_a = form.value) == null ? void 0 : _a.fields) ?? [];
    }),
    fieldsNamed: computed(() => {
      var _a;
      return ((_a = form.value) == null ? void 0 : _a.fieldsNamed) ?? [];
    }),
    fieldNames: computed(() => {
      var _a;
      return ((_a = form.value) == null ? void 0 : _a.fieldNames) ?? [];
    }),
    formData: computed(() => {
      var _a;
      return ((_a = form.value) == null ? void 0 : _a.formData) ?? {};
    }),
    errorMessages: computed(() => {
      var _a;
      return ((_a = form.value) == null ? void 0 : _a.errorMessages) || [];
    }),
    errorMessagesNamed: computed(() => {
      var _a;
      return ((_a = form.value) == null ? void 0 : _a.errorMessagesNamed) || {};
    }),
    validate: () => {
      var _a;
      return (_a = form.value) == null ? void 0 : _a.validate();
    },
    validateAsync: () => {
      var _a;
      return (_a = form.value) == null ? void 0 : _a.validateAsync();
    },
    reset: () => {
      var _a;
      return (_a = form.value) == null ? void 0 : _a.reset();
    },
    resetValidation: () => {
      var _a;
      return (_a = form.value) == null ? void 0 : _a.resetValidation();
    },
    focus: () => {
      var _a;
      return (_a = form.value) == null ? void 0 : _a.focus();
    },
    focusInvalidField: () => {
      var _a;
      return (_a = form.value) == null ? void 0 : _a.focusInvalidField();
    }
  };
};
export {
  useForm as u
};
