// vuestic-ui/dist/es/src/components/va-list/index.js
import { w as withConfigTransport } from "../../services/config-transport/withConfigTransport.js";
import { _ as _sfc_main$5 } from "./VaList.vue_vue_type_script_setup_true_lang.js";
import { _ as _sfc_main } from "./VaListItem.vue_vue_type_script_setup_true_lang.js";
import { _ as _sfc_main$1 } from "./VaListLabel.vue_vue_type_script_setup_true_lang.js";
import { _ as _sfc_main$2 } from "./VaListItemLabel.vue_vue_type_script_setup_true_lang.js";
import { _ as _sfc_main$3 } from "./VaListItemSection.vue_vue_type_script_setup_true_lang.js";
import { _ as _sfc_main$4 } from "./VaListSeparator.vue_vue_type_script_setup_true_lang.js";
const VaListItem = withConfigTransport(_sfc_main);
const VaListLabel = withConfigTransport(_sfc_main$1);
const VaListItemLabel = withConfigTransport(_sfc_main$2);
const VaListItemSection = withConfigTransport(_sfc_main$3);
const VaListSeparator = withConfigTransport(_sfc_main$4);
const VaList = withConfigTransport(_sfc_main$5);
export {
  VaListItem as V,
  VaListItemSection as a,
  VaList as b,
  VaListLabel as c,
  VaListItemLabel as d,
  VaListSeparator as e
};
