// vuestic-ui/dist/es/src/components/va-list/VaListItemLabel.vue_vue_type_script_setup_true_lang.js
import { defineComponent, computed, openBlock, createElementBlock, normalizeClass, normalizeStyle, renderSlot } from "vue";
import { u as useComponentPresetProp } from "../../composables/useComponentPreset.js";
import { u as useNumericProp } from "../../composables/useNumericProp.js";
const _sfc_main = /* @__PURE__ */ defineComponent({
  ...{
    name: "VaListItemLabel"
  },
  __name: "VaListItemLabel",
  props: {
    ...useComponentPresetProp,
    caption: { type: Boolean, default: false },
    lines: { type: [Number, String], default: 1 }
  },
  setup(__props) {
    const props = __props;
    const linesComputed = useNumericProp("lines");
    const computedClass = computed(() => ({ "va-list-item-label--caption": props.caption }));
    const computedStyle = computed(() => ({ "-webkit-line-clamp": linesComputed.value }));
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock("div", {
        class: normalizeClass(["va-list-item-label", computedClass.value]),
        style: normalizeStyle(computedStyle.value)
      }, [
        renderSlot(_ctx.$slots, "default")
      ], 6);
    };
  }
});
export {
  _sfc_main as _
};

import '../../../VaListItemLabel.css';
