// vuestic-ui/dist/web-components/main.js
import { p } from "./src/services/color/presets.js";
import { d } from "./src/services/breakpoint/index.js";
import { d as d2 } from "./src/services/global-config/types.js";
import { _ } from "./src/components/va-config/VaConfig.vue_vue_type_script_setup_true_lang.js";
import { u } from "./src/composables/useGlobalConfig.js";
import { u as u2 } from "./src/composables/useColors.js";
import { u as u3 } from "./src/composables/useBreakpoint.js";
import { u as u4 } from "./src/composables/useI18n.js";
import { u as u5 } from "./src/composables/useIcon.js";
import { u as u6 } from "./src/composables/useForm/useForm.js";
import { c } from "./src/services/icon/create-icons-config.js";
import { V } from "./src/services/icon/presets/fonts.js";
import { V as V2 } from "./src/services/icon/presets/vuestic-aliases.js";
import { G } from "./src/services/global-config/plugin/index.js";
import { C } from "./src/services/color/plugin/index.js";
import { B } from "./src/services/breakpoint/plugin/index.js";
import { V as V3 } from "./src/components/va-toast/plugin/index.js";
import { V as V4 } from "./src/components/va-dropdown/plugin/index.js";
import { V as V5 } from "./src/components/va-modal/plugin/index.js";
import { C as C2 } from "./src/services/cache/plugin/index.js";
import { C as C3 } from "./src/services/colors-classes/plugin/index.js";
import { c as c2 } from "./src/services/vue-plugin/create-vuestic/create-vuestic.js";
import { c as c3 } from "./src/services/vue-plugin/create-vuestic/create-vuestic-essential.js";
import { u as u7 } from "./src/composables/useElementTextColor.js";
import { r } from "./src/services/web-components/register-vuestic-web-components.js";
import { r as r2 } from "./src/services/web-components/register-vuestic-web-components-essential.js";
import { V as V6 } from "./src/components/va-icon/index.js";
import { V as V7 } from "./src/components/va-message-list/VaMessageListWrapper.js";
import { V as V8 } from "./src/components/va-message-list/VaMessageList.js";
import { V as V9 } from "./src/components/va-input-wrapper/index.js";
import { V as V10 } from "./src/components/va-input/index.js";
import { V as V11 } from "./src/components/va-fallback/index.js";
import { V as V12 } from "./src/components/va-collapse/index.js";
import { d as d3, b, c as c4, V as V13, a } from "./src/components/va-card/index.js";
import { V as V14 } from "./src/components/va-button/index.js";
import { b as b2, a as a2, V as V15 } from "./src/components/va-menu-list/index.js";
import { V as V16 } from "./src/components/va-accordion/index.js";
import { V as V17 } from "./src/components/va-affix/index.js";
import { V as V18 } from "./src/components/va-alert/index.js";
import { V as V19 } from "./src/components/va-app-bar/index.js";
import { V as V20 } from "./src/components/va-aspect-ratio/index.js";
import { V as V21 } from "./src/components/va-avatar/index.js";
import { V as V22 } from "./src/components/va-avatar-group/index.js";
import { V as V23 } from "./src/components/va-backtop/index.js";
import { V as V24 } from "./src/components/va-badge/index.js";
import { a as a3, V as V25 } from "./src/components/va-breadcrumbs/index.js";
import { V as V26 } from "./src/components/va-button-dropdown/index.js";
import { V as V27 } from "./src/components/va-button-group/index.js";
import { V as V28 } from "./src/components/va-button-toggle/index.js";
import { V as V29 } from "./src/components/va-carousel/index.js";
import { V as V30 } from "./src/components/va-checkbox/index.js";
import { V as V31 } from "./src/components/va-chip/index.js";
import { V as V32 } from "./src/components/va-color-indicator/index.js";
import { V as V33 } from "./src/components/va-color-input/index.js";
import { V as V34 } from "./src/components/va-color-palette/index.js";
import { V as V35 } from "./src/components/va-content/index.js";
import { V as V36 } from "./src/components/va-counter/index.js";
import { d as d4, a as a4 } from "./src/components/va-data-table/fabrics.js";
import { V as V37 } from "./src/components/va-data-table/index.js";
import { V as V38 } from "./src/components/va-date-input/index.js";
import { V as V39 } from "./src/components/va-date-picker/index.js";
import { V as V40 } from "./src/components/va-divider/index.js";
import { V as V41 } from "./src/components/va-dropdown/components/VaDropdownContent/index.js";
import { V as V42 } from "./src/components/va-dropdown/index.js";
import { V as V43 } from "./src/components/va-form/index.js";
import { V as V44 } from "./src/components/va-hover/index.js";
import { V as V45 } from "./src/components/va-image/index.js";
import { V as V46 } from "./src/components/va-infinite-scroll/index.js";
import { V as V47 } from "./src/components/va-inner-loading/index.js";
import { V as V48 } from "./src/components/va-layout/index.js";
import { b as b3, V as V49, d as d5, a as a5, c as c5, e } from "./src/components/va-list/index.js";
import { u as u8 } from "./src/components/va-modal/hooks/useModal.js";
import { V as V50 } from "./src/components/va-modal/VaModal.js";
import { V as V51, a as a6 } from "./src/components/va-navbar/index.js";
import { V as V52 } from "./src/components/va-option-list/index.js";
import { V as V53 } from "./src/components/va-pagination/index.js";
import { V as V54 } from "./src/components/va-parallax/index.js";
import { V as V55 } from "./src/components/va-popover/index.js";
import { V as V56 } from "./src/components/va-progress-bar/index.js";
import { V as V57 } from "./src/components/va-progress-circle/index.js";
import { V as V58 } from "./src/components/va-radio/index.js";
import { V as V59 } from "./src/components/va-rating/components/VaRatingItem/index.js";
import { V as V60 } from "./src/components/va-rating/index.js";
import { V as V61 } from "./src/components/va-select/components/VaSelectOptionList/index.js";
import { V as V62 } from "./src/components/va-select/components/VaSelectOption/index.js";
import { V as V63 } from "./src/components/va-select/index.js";
import { V as V64 } from "./src/components/va-separator/VaSeparator.js";
import { V as V65, a as a7 } from "./src/components/va-skeleton/index.js";
import { V as V66, a as a8, b as b4 } from "./src/components/va-sidebar/VaSidebarItem/index.js";
import { V as V67 } from "./src/components/va-sidebar/index.js";
import { V as V68 } from "./src/components/va-slider/index.js";
import { V as V69 } from "./src/components/va-spacer/VaSpacer.js";
import { V as V70 } from "./src/components/va-split/index.js";
import { V as V71, d as d6 } from "./src/components/va-stepper/index.js";
import { V as V72 } from "./src/components/va-switch/index.js";
import { V as V73 } from "./src/components/va-toast/toast.js";
import { u as u9 } from "./src/components/va-toast/hooks/useToast.js";
import { V as V74 } from "./src/components/va-tabs/components/VaTab/index.js";
import { T } from "./src/components/va-tabs/types.js";
import { V as V75 } from "./src/components/va-tabs/index.js";
import { V as V76 } from "./src/components/va-time-input/index.js";
import { V as V77 } from "./src/components/va-time-picker/index.js";
import { V as V78 } from "./src/components/va-timeline/VaTimelineItem/index.js";
import { V as V79 } from "./src/components/va-timeline/VaTimelineSeparator/index.js";
import { V as V80 } from "./src/components/va-timeline/index.js";
import { V as V81 } from "./src/components/va-tree-view/index.js";
import { V as V82 } from "./src/components/va-scroll-container/index.js";
import { V as V83 } from "./src/components/va-viewer/index.js";
import { V as V84 } from "./src/components/va-virtual-scroller/index.js";
import { V as V85 } from "./src/components/va-value/index.js";
import { V as V86 } from "./src/components/va-file-upload/VaFileUploadGalleryItem/index.js";
import { V as V87 } from "./src/components/va-file-upload/VaFileUploadList/index.js";
import { V as V88 } from "./src/components/va-file-upload/VaFileUploadListItem/index.js";
import { V as V89 } from "./src/components/va-file-upload/VaFileUploadSingleItem/index.js";
import { V as V90 } from "./src/components/va-file-upload/VaFileUploadUndo/index.js";
import { V as V91 } from "./src/components/va-file-upload/types.js";
import { V as V92 } from "./src/components/va-file-upload/index.js";
import { V as V93 } from "./src/components/va-textarea/index.js";
import { V as V94 } from "./src/components/va-menu/va-menu.js";
import { u as u10 } from "./src/components/va-menu/hooks/useMenu.js";
import { V as V95 } from "./src/components/va-form-field/index.js";
export {
  B as BreakpointConfigPlugin,
  C2 as CachePlugin,
  C as ColorConfigPlugin,
  C3 as ColorsClassesPlugin,
  G as GlobalConfigPlugin,
  T as TabsViewKey,
  V16 as VaAccordion,
  V17 as VaAffix,
  V18 as VaAlert,
  V19 as VaAppBar,
  V20 as VaAspectRatio,
  V21 as VaAvatar,
  V22 as VaAvatarGroup,
  V23 as VaBacktop,
  V24 as VaBadge,
  a3 as VaBreadcrumbs,
  V25 as VaBreadcrumbsItem,
  V14 as VaButton,
  V26 as VaButtonDropdown,
  V27 as VaButtonGroup,
  V28 as VaButtonToggle,
  d3 as VaCard,
  b as VaCardActions,
  c4 as VaCardBlock,
  V13 as VaCardContent,
  a as VaCardTitle,
  V29 as VaCarousel,
  V30 as VaCheckbox,
  V31 as VaChip,
  V12 as VaCollapse,
  V32 as VaColorIndicator,
  V33 as VaColorInput,
  V34 as VaColorPalette,
  _ as VaConfig,
  V35 as VaContent,
  V36 as VaCounter,
  V37 as VaDataTable,
  V38 as VaDateInput,
  V39 as VaDatePicker,
  V40 as VaDivider,
  V42 as VaDropdown,
  V41 as VaDropdownContent,
  V4 as VaDropdownPlugin,
  V11 as VaFallback,
  V92 as VaFileUpload,
  V86 as VaFileUploadGalleryItem,
  V91 as VaFileUploadKey,
  V87 as VaFileUploadList,
  V88 as VaFileUploadListItem,
  V89 as VaFileUploadSingleItem,
  V90 as VaFileUploadUndo,
  V43 as VaForm,
  V95 as VaFormField,
  V44 as VaHover,
  V6 as VaIcon,
  V45 as VaImage,
  V46 as VaInfiniteScroll,
  V47 as VaInnerLoading,
  V10 as VaInput,
  V9 as VaInputWrapper,
  V48 as VaLayout,
  b3 as VaList,
  V49 as VaListItem,
  d5 as VaListItemLabel,
  a5 as VaListItemSection,
  c5 as VaListLabel,
  e as VaListSeparator,
  V94 as VaMenu,
  b2 as VaMenuGroup,
  a2 as VaMenuItem,
  V15 as VaMenuList,
  V8 as VaMessageList,
  V7 as VaMessageListWrapper,
  V50 as VaModal,
  V5 as VaModalPlugin,
  V51 as VaNavbar,
  a6 as VaNavbarItem,
  V52 as VaOptionList,
  V53 as VaPagination,
  V54 as VaParallax,
  V55 as VaPopover,
  V56 as VaProgressBar,
  V57 as VaProgressCircle,
  V58 as VaRadio,
  V60 as VaRating,
  V59 as VaRatingItem,
  V82 as VaScrollContainer,
  V63 as VaSelect,
  V62 as VaSelectOption,
  V61 as VaSelectOptionList,
  V64 as VaSeparator,
  V67 as VaSidebar,
  V66 as VaSidebarItem,
  a8 as VaSidebarItemContent,
  b4 as VaSidebarItemTitle,
  V65 as VaSkeleton,
  a7 as VaSkeletonGroup,
  V68 as VaSlider,
  V69 as VaSpacer,
  V70 as VaSplit,
  V71 as VaStepper,
  V72 as VaSwitch,
  V74 as VaTab,
  V75 as VaTabs,
  V93 as VaTextarea,
  V76 as VaTimeInput,
  V77 as VaTimePicker,
  V80 as VaTimeline,
  V78 as VaTimelineItem,
  V79 as VaTimelineSeparator,
  V73 as VaToast,
  V3 as VaToastPlugin,
  V81 as VaTreeView,
  V85 as VaValue,
  V83 as VaViewer,
  V84 as VaVirtualScroller,
  V2 as VuesticIconAliases,
  V as VuesticIconFonts,
  p as colorsPreset,
  c as createIconsConfig,
  c2 as createVuestic,
  c3 as createVuesticEssential,
  d4 as defineVaDataTableColumns,
  a4 as defineVaDataTableItems,
  d6 as defineVaStepperSteps,
  d2 as defineVuesticConfig,
  r as registerVuesticWebComponents,
  r2 as registerVuesticWebComponentsEssential,
  d as thresholdsPreset,
  u3 as useBreakpoint,
  u2 as useColors,
  u7 as useElementTextColor,
  u6 as useForm,
  u as useGlobalConfig,
  u4 as useI18nConfig,
  u5 as useIcons,
  u10 as useMenu,
  u8 as useModal,
  u9 as useToast
};
