// vuestic-ui/dist/web-components/src/components/va-file-upload/VaFileUploadSingleItem/VaFileUploadSingleItem.vue_vue_type_script_setup_true_lang.js
import { defineComponent, openBlock, createBlock, unref, withCtx, createVNode, createElementVNode, toDisplayString, createTextVNode, createCommentVNode } from "vue";
import { V as VaFileUploadKey } from "../types.js";
import { u as useTranslation } from "../../../composables/useTranslation.js";
import { u as useStrictInject } from "../../../composables/useStrictInject.js";
import { V as VaListItem, a as VaListItemSection } from "../../va-list/index.js";
import { V as VaButton } from "../../va-button/index.js";
const _hoisted_1 = { class: "va-file-upload-single-item__name" };
const INJECTION_ERROR_MESSAGE = "The VaFileUploadSingleItem component should be used in the context of VaFileUpload component";
const _sfc_main = /* @__PURE__ */ defineComponent({
  ...{
    name: "VaFileUploadSingleItem"
  },
  __name: "VaFileUploadSingleItem",
  props: {
    file: { type: Object, default: null },
    ariaRemoveFileLabel: { type: String, default: "$t:removeFile" }
  },
  emits: ["remove"],
  setup(__props, { emit: __emit }) {
    const { t, tp } = useTranslation();
    const { disabled } = useStrictInject(VaFileUploadKey, INJECTION_ERROR_MESSAGE);
    return (_ctx, _cache) => {
      return openBlock(), createBlock(unref(VaListItem), {
        disabled: unref(disabled),
        "aria-disabled": unref(disabled),
        class: "va-file-upload-single-item",
        tabindex: "-1"
      }, {
        default: withCtx(() => [
          createVNode(unref(VaListItemSection), { class: "va-file-upload-single-item__content" }, {
            default: withCtx(() => [
              createElementVNode("div", _hoisted_1, toDisplayString(__props.file && __props.file.name), 1),
              !unref(disabled) ? (openBlock(), createBlock(unref(VaButton), {
                key: 0,
                class: "va-file-upload-single-item__button",
                "aria-label": unref(tp)(_ctx.$props.ariaRemoveFileLabel),
                size: "small",
                color: "danger",
                preset: "secondary",
                onClick: _cache[0] || (_cache[0] = ($event) => _ctx.$emit("remove"))
              }, {
                default: withCtx(() => [
                  createTextVNode(" Delete ")
                ]),
                _: 1
              }, 8, ["aria-label"])) : createCommentVNode("", true)
            ]),
            _: 1
          })
        ]),
        _: 1
      }, 8, ["disabled", "aria-disabled"]);
    };
  }
});
export {
  _sfc_main as _
};
