// vuestic-ui/dist/web-components/src/components/va-date-picker/components/VaDatePickerCell.vue_vue_type_script_setup_true_lang.js
import { defineComponent, useCssVars, unref, computed, openBlock, createElementBlock, normalizeClass, withKeys, withModifiers, renderSlot } from "vue";
import { u as useColors } from "../../../composables/useColors.js";
import { u as useTextColor } from "../../../composables/useTextColor.js";
const _hoisted_1 = {
  key: 0,
  class: "va-date-picker-cell va-date-picker-cell_clear"
};
const _hoisted_2 = ["onKeypress"];
const _sfc_main = /* @__PURE__ */ defineComponent({
  ...{
    name: "VaDatePickerCell"
  },
  __name: "VaDatePickerCell",
  props: {
    otherMonth: { type: Boolean, default: false },
    today: { type: Boolean, default: false },
    inRange: { type: Boolean, default: false },
    disabled: { type: Boolean, default: false },
    selected: { type: Boolean, default: false },
    weekend: { type: Boolean, default: false },
    hidden: { type: Boolean, default: false },
    focused: { type: Boolean, default: false },
    highlightWeekend: { type: Boolean, default: false },
    highlightToday: { type: Boolean, default: false },
    readonly: { type: Boolean, default: false },
    color: { type: String, default: "primary" }
  },
  emits: ["click"],
  setup(__props, { emit: __emit }) {
    useCssVars((_ctx) => ({
      "50437be3": bg.value,
      "be8fc7f6": unref(textColorComputed)
    }));
    const props = __props;
    const emit = __emit;
    const onClick = () => {
      if (!props.disabled) {
        emit("click");
      }
    };
    const { getColor } = useColors();
    const bg = computed(() => getColor(props.color));
    const { textColorComputed } = useTextColor(bg);
    return (_ctx, _cache) => {
      return __props.hidden ? (openBlock(), createElementBlock("div", _hoisted_1)) : (openBlock(), createElementBlock("div", {
        key: 1,
        class: normalizeClass(["va-date-picker-cell", {
          "va-date-picker-cell_other-month": __props.otherMonth,
          "va-date-picker-cell_today": __props.highlightToday && __props.today,
          "va-date-picker-cell_in-range": __props.inRange,
          "va-date-picker-cell_disabled": __props.disabled,
          "va-date-picker-cell_highlighted-weekend": __props.highlightWeekend && __props.weekend,
          "va-date-picker-cell_selected": __props.selected,
          "va-date-picker-cell_focused": __props.focused,
          "va-date-picker-cell_readonly": __props.readonly
        }]),
        onClick,
        onKeypress: withKeys(withModifiers(onClick, ["prevent", "stop"]), ["space", "enter"])
      }, [
        renderSlot(_ctx.$slots, "default")
      ], 42, _hoisted_2));
    };
  }
});
export {
  _sfc_main as _
};
