// vuestic-ui/dist/web-components/src/components/va-virtual-scroller/VaVirtualScroller.vue_vue_type_script_setup_true_lang.js
import { defineComponent, ref, computed, watch, openBlock, createElementBlock, normalizeClass, unref, normalizeStyle, renderSlot, normalizeProps, guardReactiveProps, createElementVNode, Fragment, renderList, mergeProps } from "vue";
import { p as pick } from "../../../vendor.js";
import { u as useVirtualScrollerSizesProps, a as useVirtualScrollerSizes } from "./useVirtualScrollerSizes.js";
import { u as useTrackByProps, a as useTrackBy } from "../../composables/useTrackBy.js";
import { u as useNumericProp } from "../../composables/useNumericProp.js";
import { u as useEvent } from "../../composables/useEvent.js";
import { u as useBem } from "../../composables/useBem.js";
const _sfc_main = /* @__PURE__ */ defineComponent({
  ...{
    name: "VaVirtualScroller"
  },
  __name: "VaVirtualScroller",
  props: {
    ...useTrackByProps,
    ...useVirtualScrollerSizesProps,
    items: { type: Array, default: () => [] },
    bench: { type: [Number, String], default: 10, validator: (v) => Number(v) >= 0 },
    disabled: { type: Boolean, default: false },
    table: { type: Boolean, default: false }
  },
  emits: ["scroll:bottom"],
  setup(__props, { expose: __expose, emit: __emit }) {
    const props = __props;
    const emit = __emit;
    const listScrollPosition = ref(0);
    const benchComputed = useNumericProp("bench");
    const scrollDirection = computed(() => props.horizontal ? "scrollLeft" : "scrollTop");
    const handleScroll = () => {
      if (!wrapper.value) {
        return;
      }
      listScrollPosition.value = wrapper.value[scrollDirection.value];
    };
    if (!props.disabled) {
      useEvent("scroll", handleScroll, true);
    }
    const { list, wrapper, itemSize, wrapperSize } = useVirtualScrollerSizes(props, listScrollPosition);
    const { getKey } = useTrackBy(props);
    const uniqueKey = (item, index, defaultValue) => getKey(item, index, defaultValue);
    watch(listScrollPosition, (newValue) => {
      if (newValue + wrapperSize.value === containerSize.value) {
        emit("scroll:bottom");
      }
    });
    const renderStartIndex = computed(() => {
      return Math.max(0, Math.floor(listScrollPosition.value / itemSize.value) - benchComputed.value);
    });
    const renderItemsAmount = computed(() => {
      var _a;
      if (!((_a = props.items) == null ? void 0 : _a.length)) {
        return 0;
      }
      return props.disabled ? props.items.length : Math.min(props.items.length - renderStartIndex.value, Math.ceil(wrapperSize.value / itemSize.value) + benchComputed.value * 2);
    });
    const renderEndIndex = computed(() => renderStartIndex.value + renderItemsAmount.value);
    const renderBuffer = computed(() => {
      var _a;
      if (!((_a = props.items) == null ? void 0 : _a.length)) {
        return [];
      }
      return props.items.slice(renderStartIndex.value, renderEndIndex.value);
    });
    const sizeAttribute = computed(() => props.horizontal ? "width" : "height");
    const isDisabledVirtualTable = computed(() => props.table && props.disabled);
    const wrapperStyleComputed = computed(() => ({
      [sizeAttribute.value]: isDisabledVirtualTable.value || !wrapperSize.value ? void 0 : `${wrapperSize.value}px`
    }));
    const wrapperClassComputed = useBem("va-virtual-scroller", () => ({
      ...pick(props, ["horizontal"])
    }));
    const containerSize = computed(() => {
      var _a;
      return (((_a = props.items) == null ? void 0 : _a.length) ?? 0) * itemSize.value;
    });
    const containerStyleComputed = computed(() => ({
      [sizeAttribute.value]: isDisabledVirtualTable.value ? void 0 : `${containerSize.value}px`
    }));
    const currentListOffset = computed(() => renderStartIndex.value * itemSize.value);
    const listStyleComputed = computed(() => ({
      transform: `translate${props.horizontal ? "X" : "Y"}(${currentListOffset.value}px)`
    }));
    const scrollToAttribute = computed(() => props.horizontal ? "left" : "top");
    const virtualScrollTo = (index) => {
      var _a;
      if (!index && index !== 0) {
        return;
      }
      (_a = wrapper.value) == null ? void 0 : _a.scrollTo({ [scrollToAttribute.value]: index * itemSize.value });
    };
    __expose({
      scrollToAttribute,
      virtualScrollTo
    });
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock("div", {
        ref_key: "wrapper",
        ref: wrapper,
        class: normalizeClass(["va-virtual-scroller", unref(wrapperClassComputed)]),
        style: normalizeStyle(wrapperStyleComputed.value)
      }, [
        renderSlot(_ctx.$slots, "content", normalizeProps(guardReactiveProps({ containerStyleComputed: containerStyleComputed.value, listStyleComputed: listStyleComputed.value, renderBuffer: renderBuffer.value, uniqueKey, currentListOffset: currentListOffset.value })), () => [
          createElementVNode("div", {
            class: "va-virtual-scroller__container",
            style: normalizeStyle(containerStyleComputed.value)
          }, [
            createElementVNode("div", {
              ref_key: "list",
              ref: list,
              role: "list",
              class: "va-virtual-scroller__list",
              style: normalizeStyle(listStyleComputed.value)
            }, [
              (openBlock(true), createElementBlock(Fragment, null, renderList(renderBuffer.value, (item, index) => {
                return renderSlot(_ctx.$slots, "default", normalizeProps(mergeProps({
                  key: uniqueKey(item, index)
                }, { item, index: renderStartIndex.value + index })));
              }), 128))
            ], 4)
          ], 4)
        ])
      ], 6);
    };
  }
});
export {
  _sfc_main as _
};
