// vuestic-ui/dist/web-components/src/components/va-menu-list/components/VaMenuGroup.vue_vue_type_script_setup_true_lang.js
import { defineComponent, useCssVars, computed, openBlock, createElementBlock, Fragment, createElementVNode, toDisplayString, renderSlot } from "vue";
import { u as useColors } from "../../../composables/useColors.js";
const _hoisted_1 = {
  class: "va-menu-list__group-name-wrapper",
  colspan: "99999"
};
const _hoisted_2 = { class: "va-menu-list__group-name" };
const _sfc_main = /* @__PURE__ */ defineComponent({
  ...{
    name: "VaMenuGroup"
  },
  __name: "VaMenuGroup",
  props: {
    groupName: {
      type: String,
      required: true
    },
    color: {
      type: String,
      default: "secondary"
    }
  },
  setup(__props) {
    useCssVars((_ctx) => ({
      "4731859b": colorComputed.value
    }));
    const props = __props;
    const { getColor } = useColors();
    const colorComputed = computed(() => getColor(props.color));
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock(Fragment, null, [
        createElementVNode("div", _hoisted_1, [
          createElementVNode("span", _hoisted_2, toDisplayString(__props.groupName), 1)
        ]),
        renderSlot(_ctx.$slots, "default", {}, void 0, true)
      ], 64);
    };
  }
});
export {
  _sfc_main as _
};
