// vuestic-ui/dist/web-components/src/components/va-input/hooks/useCleave.js
import { ref, computed, watchEffect, onBeforeUnmount } from "vue";
import Cleave from "cleave.js";
const DEFAULT_MASK_TOKENS = {
  creditCard: {
    creditCard: true
  },
  date: {
    date: true,
    datePattern: ["d", "m", "Y"]
  },
  time: {
    time: true,
    timePattern: ["h", "m"],
    timeFormat: "24"
  },
  numeral: {
    numeral: true,
    numeralThousandsGroupStyle: "thousand"
  }
};
const useCleaveProps = {
  mask: { type: [String, Object], default: "" },
  returnRaw: { type: Boolean, default: true }
};
const useCleave = (element, props, syncValue) => {
  const cleave = ref();
  const getMask = (mask2) => {
    if (typeof mask2 === "string") {
      return DEFAULT_MASK_TOKENS[mask2] ? { ...DEFAULT_MASK_TOKENS[mask2] } : null;
    }
    return { ...mask2 };
  };
  const destroyCleave = () => {
    if (cleave.value) {
      cleave.value.destroy();
    }
  };
  const mask = computed(() => getMask(props.mask));
  const cleaveEnabled = computed(() => {
    return mask.value && Object.keys(mask.value).length;
  });
  watchEffect(() => {
    destroyCleave();
    if (!element.value) {
      return;
    }
    if (!cleaveEnabled.value || !mask.value) {
      return;
    }
    cleave.value = new Cleave(element.value, mask.value);
    cleave.value.properties.onValueChanged = ({ target: { rawValue, value } }) => {
      if (props.returnRaw) {
        syncValue.value = rawValue;
      } else {
        syncValue.value = value;
      }
    };
  });
  onBeforeUnmount(() => {
    destroyCleave();
  });
  const computedValue = computed(() => {
    if (cleave.value) {
      if (props.returnRaw && syncValue.value === cleave.value.getRawValue()) {
        return cleave.value.getFormattedValue();
      }
    }
    return syncValue.value;
  });
  const onInput = (event) => {
    const value = event.target.value;
    if (!cleaveEnabled.value) {
      syncValue.value = value;
    }
  };
  return {
    cleave,
    cleaveEnabled,
    computedValue,
    onInput
  };
};
export {
  useCleave as a,
  useCleaveProps as u
};
